src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,447 @@
+  elseif crs_type <: LatLonAlt
+    crs_type(buff[2], buff[1], buff[3])  # LatLonAlt(lat, lon, alt)
+  else
+    crs_type(buff...)  # fallback for other coordinate systems
+  end
+end
+
+function parse_wkb_point(io::IOBuffer, parse_coords::Function, crs_type)
+  buff = parse_coords(io)
+  # Validate coordinates
+  if isnan(buff[1]) || isnan(buff[2])
+    error("Invalid coordinates: NaN values found")
+  end
+  coord = create_crs_coord(buff, crs_type)
+  return Point(coord)
+end
+
+# Batch parse multiple points efficiently
+function parse_wkb_points(io::IOBuffer, n::Int, parse_coords::Function, crs_type::Type{<:CRS})
+  # Read n coordinate buffers and convert to points
+  map(1:n) do _
+    buff = parse_coords(io)
+    coord = create_crs_coord(buff, crs_type)
+    Point(coord)
+  end
+end
+
+function parse_wkb_linestring(io::IOBuffer, parse_coords::Function, crs_type; is_ring::Bool=false)
+  num_points = Int(ltoh(read(io, UInt32)))
+  
+  # Use batch processing
+  points = parse_wkb_points(io, num_points, parse_coords, crs_type)
+  
+  # Check if WKB writer included duplicate closing point (varies by implementation)
+  has_equal_ends = length(points) >= 2 && points[1] == points[end]
+  
+  # Two-flag logic explained:
+  # - is_ring: Semantic flag indicating this linestring is part of a polygon structure
+  # - has_equal_ends: WKB format flag indicating if writer included duplicate closing point
+  if is_ring || has_equal_ends
+    # Return a Ring, exclude duplicate last point if present
+    if has_equal_ends
+      return Ring(points[1:end-1]...)
+    else
+      return Ring(points...)
+    end
+  else
+    # It's an open linestring, return a Rope
+    return Rope(points...)
+  end
+end
+
+function parse_wkb_polygon(io::IOBuffer, parse_coords::Function, crs_type)
+  num_rings = Int(ltoh(read(io, UInt32)))
+  
+  # Parse exterior ring using ring parsing function
+  exterior_ring = parse_wkb_linestring(io, parse_coords, crs_type; is_ring=true)
+  
+  # Parse interior rings (holes) using map-do syntax
+  hole_rings = map(2:num_rings) do ring_idx
+    parse_wkb_linestring(io, parse_coords, crs_type; is_ring=true)
+  end
+  
+  return PolyArea(exterior_ring, hole_rings...)
+end
+
+
+function parse_wkb_multipoint(io::IOBuffer, parse_coords::Function, crs_type)
+  num_points = Int(ltoh(read(io, UInt32)))
+  
+  points = map(1:num_points) do _
+    parse_wkb_geometry_from_io(io, crs_type)
+  end
+  
+  return Multi(points)
+end
+
+function parse_wkb_multilinestring(io::IOBuffer, parse_coords::Function, crs_type)
+  num_lines = Int(ltoh(read(io, UInt32)))
+  
+  lines = map(1:num_lines) do _
+    parse_wkb_geometry_from_io(io, crs_type)
+  end
+  
+  return Multi(lines)
+end
+
+function parse_wkb_multipolygon(io::IOBuffer, parse_coords::Function, crs_type)
+  num_polygons = Int(ltoh(read(io, UInt32)))
+  
+  polygons = map(1:num_polygons) do _
+    parse_wkb_geometry_from_io(io, crs_type)
```

Comment:
If you are parsing a multipolygon, then you should use a parse_wkb_polygon in the loop.